src/dmd/expressionsem.d

```diff
@@ -10653,10 +10673,16 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             // For e1 and e2 of struct type, lowers e1 == e2 to object.__equals(e1, e2)
             // and e1 != e2 to !(object.__equals(e1, e2)).
 
+            // Verify that the hook `object.__equals` exist
+            if (!verifyHookExist(exp.loc, sc, Id.__equals, "equal checks on arrays"))
+            {
+                result = new ErrorExp();
+                return;
+            }
+
             Expression __equals = new IdentifierExp(exp.loc, Id.empty);
-            Identifier id = Identifier.idPool("__equals");
             __equals = new DotIdExp(exp.loc, __equals, Id.object);
-            __equals = new DotIdExp(exp.loc, __equals, id);
+            __equals = new DotIdExp(exp.loc, __equals, Id.__equals);
```

Comment:
It's best to put things like this in a separate PR.

Reply:
Will do